clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -338,6 +383,175 @@ impl FunctionType {
     }
 }
 
+pub fn trait_check_trait_compliance(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    actual_trait_identifier: &TraitIdentifier,
+    actual_trait: &BTreeMap<ClarityName, FunctionSignature>,
+    expected_trait_identifier: &TraitIdentifier,
+    expected_trait: &BTreeMap<ClarityName, FunctionSignature>,
+) -> CheckResult<()> {
+    // Shortcut for the simple case when the two traits are the same.
+    if actual_trait_identifier == expected_trait_identifier {
+        return Ok(());
+    }
+
+    for (func_name, expected_sig) in expected_trait.iter() {
+        if let Some(func) = actual_trait.get(func_name) {
+            let args_iter = expected_sig.args.iter().zip(func.args.iter());
+            for (expected_type, actual_type) in args_iter {
+                if inner_type_check_type(db, contract_context, actual_type, expected_type).is_err()
+                {
+                    return Err(CheckErrors::IncompatibleTrait(
+                        expected_trait_identifier.clone(),
+                        actual_trait_identifier.clone(),
+                    )
+                    .into());
+                }
+            }
+            if inner_type_check_type(db, contract_context, &func.returns, &expected_sig.returns)
+                .is_err()
+            {
+                return Err(CheckErrors::IncompatibleTrait(
+                    expected_trait_identifier.clone(),
+                    actual_trait_identifier.clone(),
+                )
+                .into());
+            }
+        } else {
+            return Err(CheckErrors::IncompatibleTrait(
+                expected_trait_identifier.clone(),
+                actual_trait_identifier.clone(),
+            )
+            .into());
+        }
+    }
+    Ok(())
+}
+
+fn inner_type_check_type(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    actual_type: &TypeSignature,
+    expected_type: &TypeSignature,
+) -> TypeResult {
+    // Recurse into values to check embedded traits properly
+    match (actual_type, expected_type) {
+        (
+            TypeSignature::OptionalType(atom_inner_type),
+            TypeSignature::OptionalType(expected_inner_type),
+        ) => {
+            inner_type_check_type(db, contract_context, atom_inner_type, expected_inner_type)?;
```

Comment:
This looks like a stack overflow bug waiting to happen.  Can you bind the recursion depth?

Reply:
My thought was that those `TypeSignature`s cannot be created without triggering an error in the case that they are too deep, but it doesn't hurt to add a check here.

Reply:
Added in 7aff8b5a8